File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.Phasors/Ieee1344/CommonFrameHeader.vb
'*******************************************************************************************************
'  CommonFrameHeader.vb - IEEE 1344 Common frame header functions
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  11/12/2004 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.IO
Imports Tva.DateTime
Imports Tva.Phasors.Common
Imports Tva.Phasors.Ieee1344.Common
Imports Tva.IO.Compression.Common

Namespace Ieee1344

    ' This class generates and parses a frame header specfic to C37.118
    <CLSCompliant(False)> _
    Public NotInheritable Class CommonFrameHeader

#Region " Internal Common Frame Header Instance Class "

        ' This class is used to temporarily hold parsed frame header
        Friend Class CommonFrameHeaderInstance

            Implements ICommonFrameHeader

            Private m_idCode As UInt64
            Private m_sampleCount As Int16
            Private m_statusFlags As Int16
            Private m_ticks As Long
            Private m_frameQueue As MemoryStream

            Public ReadOnly Property This() As IChannel Implements IChannel.This
                Get
                    Return Me
                End Get
            End Property

            Private ReadOnly Property IFrameThis() As Measurements.IFrame Implements Measurements.IFrame.This
                Get
                    Return Me
                End Get
            End Property

            Public ReadOnly Property InheritedType() As System.Type Implements IChannel.InheritedType
                Get
                    Return Me.GetType()
                End Get
            End Property

            Public Property IDCode() As UInt64 Implements ICommonFrameHeader.IDCode
                Get
                    Return m_idCode
                End Get
                Set(ByVal value As UInt64)
                    m_idCode = value
                End Set
            End Property

            Private Property IChannelFrameIDCode() As UInt16 Implements IChannelFrame.IDCode
                Get
                    If m_idCode > UInt16.MaxValue Then
                        Return UInt16.MaxValue
                    Else
                        Return Convert.ToUInt16(m_idCode)
                    End If
                End Get
                Set(ByVal value As UInt16)
                    m_idCode = value
                End Set
            End Property

            Public Sub AppendFrameImage(ByVal binaryImage As Byte(), ByVal offset As Integer, ByVal length As Integer)

                ' Validate CRC of frame image being appended
                If Not ChecksumIsValid(binaryImage, offset, length) Then
                    m_frameQueue = Nothing
                    Throw New InvalidOperationException("Invalid binary image detected - check sum of individual IEEE 1344 interleaved configuration or header frame did not match")
                End If

                ' Create new frame queue to hold combined binary image, if it doesn't already exist
                If m_frameQueue Is Nothing Then
                    m_frameQueue = New MemoryStream

                    ' Include initial header in new stream...
                    m_frameQueue.Write(binaryImage, offset, CommonFrameHeader.BinaryLength)
                End If

                ' Skip past header
                offset += CommonFrameHeader.BinaryLength

                ' Include frame image
                m_frameQueue.Write(binaryImage, offset, length - CommonFrameHeader.BinaryLength)

            End Sub

            Private Function ChecksumIsValid(ByVal buffer As Byte(), ByVal startIndex As Int32, ByVal length As Int32) As Boolean

                Dim sumLength As Int16 = length - 2
                Return EndianOrder.BigEndian.ToUInt16(buffer, startIndex + sumLength) = CRC16(UInt16.MaxValue, buffer, startIndex, sumLength)

            End Function

            Public Property IsFirstFrame() As Boolean
                Get
                    Return CommonFrameHeader.IsFirstFrame(Me)
                End Get
                Set(ByVal value As Boolean)
                    CommonFrameHeader.IsFirstFrame(Me) = value
                End Set
            End Property

            Public Property IsLastFrame() As Boolean
                Get
                    Return CommonFrameHeader.IsLastFrame(Me)
                End Get
                Set(ByVal value As Boolean)
                    CommonFrameHeader.IsLastFrame(Me) = value
                End Set
            End Property

            Public Property FrameCount() As Int16
                Get
                    Return CommonFrameHeader.FrameCount(Me)
                End Get
                Set(ByVal value As Int16)
                    CommonFrameHeader.FrameCount(Me) = value
                End Set
            End Property

            Public ReadOnly Property FrameType() As FrameType Implements ICommonFrameHeader.FrameType
                Get
                    Return CommonFrameHeader.FrameType(Me)
                End Get
            End Property

            Public ReadOnly Property FrameLength() As Int16 Implements ICommonFrameHeader.FrameLength
                Get
                    If m_frameQueue IsNot Nothing Then
                        ' If we are cumulating frames, we use this total length instead of length parsed from individual frame
                        Return BinaryLength
                    Else
                        Return CommonFrameHeader.FrameLength(Me)
                    End If
                End Get
            End Property

            Public ReadOnly Property DataLength() As Int16 Implements ICommonFrameHeader.DataLength
                Get
                    Return CommonFrameHeader.DataLength(Me)
                End Get
            End Property

            Public Property SynchronizationIsValid() As Boolean Implements ICommonFrameHeader.SynchronizationIsValid
                Get
                    Return CommonFrameHeader.SynchronizationIsValid(Me)
                End Get
                Set(ByVal value As Boolean)
                    CommonFrameHeader.SynchronizationIsValid(Me) = value
                End Set
            End Property

            Public Property DataIsValid() As Boolean Implements ICommonFrameHeader.DataIsValid
                Get
                    Return CommonFrameHeader.DataIsValid(Me)
                End Get
                Set(ByVal value As Boolean)
                    CommonFrameHeader.DataIsValid(Me) = value
                End Set
            End Property

            Public Property TriggerStatus() As TriggerStatus Implements ICommonFrameHeader.TriggerStatus
                Get
                    Return CommonFrameHeader.TriggerStatus(Me)
                End Get
                Set(ByVal value As TriggerStatus)
                    CommonFrameHeader.TriggerStatus(Me) = value
                End Set
            End Property

            Public Property InternalSampleCount() As Int16 Implements ICommonFrameHeader.InternalSampleCount
                Get
                    Return m_sampleCount
                End Get
                Set(ByVal value As Int16)
                    m_sampleCount = value
                End Set
            End Property

            Public Property InternalStatusFlags() As Int16 Implements ICommonFrameHeader.InternalStatusFlags
                Get
                    Return m_statusFlags
                End Get
                Set(ByVal value As Int16)
                    m_statusFlags = value
                End Set
            End Property

            Public Property Ticks() As Long Implements IChannelFrame.Ticks
                Get
                    Return m_ticks
                End Get
                Set(ByVal value As Long)
                    m_ticks = value
                End Set
            End Property

            Public ReadOnly Property TimeTag() As NtpTimeTag Implements ICommonFrameHeader.TimeTag
                Get
                    Return CommonFrameHeader.TimeTag(Me)
                End Get
            End Property

            Private ReadOnly Property IChannelFrameTimeTag() As UnixTimeTag Implements IChannelFrame.TimeTag
                Get
                    Return New UnixTimeTag(Timestamp)
                End Get
            End Property

            Public ReadOnly Property BinaryImage() As Byte() Implements IChannel.BinaryImage
                Get
                    If m_frameQueue Is Nothing Then
                        Return Nothing
                    Else
                        Return m_frameQueue.ToArray()
                    End If
                End Get
            End Property

            Public ReadOnly Property BinaryLength() As UInt16 Implements IChannel.BinaryLength
                Get
                    If m_frameQueue Is Nothing Then
                        Return 0
                    Else
                        Return m_frameQueue.Length
                    End If
                End Get
            End Property

            Public Sub ParseBinaryImage(ByVal state As IChannelParsingState, ByVal binaryImage() As Byte, ByVal startIndex As Integer) Implements IChannel.ParseBinaryImage

                Throw New NotImplementedException()

            End Sub

            Public ReadOnly Property Cells() As Object Implements IChannelFrame.Cells
                Get
                    Return Nothing
                End Get
            End Property

            Private ReadOnly Property IChannelFrameFrameType() As FundamentalFrameType Implements IChannelFrame.FrameType
                Get
                    ' Translate IEEE 1344 specific frame type to fundamental frame type
                    Select Case FrameType
                        Case Ieee1344.FrameType.DataFrame
                            Return FundamentalFrameType.DataFrame
                        Case Ieee1344.FrameType.ConfigurationFrame
                            Return FundamentalFrameType.ConfigurationFrame
                        Case Ieee1344.FrameType.HeaderFrame
                            Return FundamentalFrameType.HeaderFrame
                        Case Else
                            Return FundamentalFrameType.Undetermined
                    End Select
                End Get
            End Property

            Public Property Published() As Boolean Implements IChannelFrame.Published
                Get
                    Return False
                End Get
                Set(ByVal value As Boolean)
                    Throw New NotImplementedException()
                End Set
            End Property

            ' This frame is not complete - it only represents the parsed common "header" for frames
            Public ReadOnly Property IsPartial() As Boolean Implements IChannelFrame.IsPartial
                Get
                    Return True
                End Get
            End Property

            Public ReadOnly Property Timestamp() As Date Implements IChannelFrame.Timestamp
                Get
                    Return New Date(m_ticks)
                End Get
            End Property

            Public Function CompareTo(ByVal obj As Object) As Integer Implements System.IComparable.CompareTo

                If TypeOf obj Is IChannelFrame Then
                    Return m_ticks.CompareTo(DirectCast(obj, IChannelFrame).Ticks)
                Else
                    Throw New ArgumentException(InheritedType.Name & " can only be compared with other IChannelFrames...")
                End If
            End Function

            Private ReadOnly Property IFrameMeasurements() As System.Collections.Generic.IDictionary(Of Integer, Measurements.IMeasurement) Implements Measurements.IFrame.Measurements
                Get
                    Throw New NotImplementedException()
                End Get
            End Property

        End Class

#End Region

        Public Const BinaryLength As UInt16 = 8

        Private Sub New()

            ' This class contains only global functions and is not meant to be instantiated

        End Sub

        Public Shared Function ParseBinaryImage(ByVal configurationFrame As ConfigurationFrame, ByVal binaryImage As Byte(), ByVal startIndex As Int32) As ICommonFrameHeader

            With New CommonFrameHeaderInstance
                Dim secondOfCentury As UInt32

                secondOfCentury = EndianOrder.BigEndian.ToUInt32(binaryImage, startIndex)
                .InternalSampleCount = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 4)
                .InternalStatusFlags = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 6)

                If .FrameType = Ieee1344.FrameType.DataFrame AndAlso configurationFrame IsNot Nothing Then
                    ' Data frames have subsecond time information
                    .Ticks = (New NtpTimeTag(secondOfCentury + SampleCount(.This) / System.Math.Floor(MaximumSampleCount / configurationFrame.Period) / configurationFrame.FrameRate)).ToDateTime.Ticks
                Else
                    ' For other frames, the best timestamp you can get is down to the whole second
                    .Ticks = (New NtpTimeTag(secondOfCentury)).ToDateTime.Ticks
                End If

                Return .This
            End With

        End Function

        Public Shared Function BinaryImage(ByVal frameHeader As ICommonFrameHeader) As Byte()

            Dim buffer As Byte() = CreateArray(Of Byte)(BinaryLength)

            ' Make sure frame length gets included in status flags for generated binary image
            FrameLength(frameHeader) = frameHeader.BinaryLength

            With frameHeader
                EndianOrder.BigEndian.CopyBytes(Convert.ToUInt32(.TimeTag.Value), buffer, 0)
                EndianOrder.BigEndian.CopyBytes(.InternalSampleCount, buffer, 4)
                EndianOrder.BigEndian.CopyBytes(.InternalStatusFlags, buffer, 6)
            End With

            Return buffer

        End Function

        Public Shared Sub Clone(ByVal sourceFrameHeader As ICommonFrameHeader, ByVal destinationFrameHeader As ICommonFrameHeader)

            With destinationFrameHeader
                .Ticks = sourceFrameHeader.Ticks
                .InternalSampleCount = sourceFrameHeader.InternalSampleCount
                .InternalStatusFlags = sourceFrameHeader.InternalStatusFlags
            End With

        End Sub

        Public Shared ReadOnly Property TimeTag(ByVal frameHeader As ICommonFrameHeader) As NtpTimeTag
            Get
                Return New NtpTimeTag(New Date(frameHeader.Ticks))
            End Get
        End Property

        Public Shared Property FrameType(ByVal frameHeader As ICommonFrameHeader) As FrameType
            Get
                Return frameHeader.InternalSampleCount And FrameTypeMask
            End Get
            Friend Set(ByVal value As FrameType)
                frameHeader.InternalSampleCount = (frameHeader.InternalSampleCount And Not FrameTypeMask) Or value
            End Set
        End Property

        Public Shared Property SampleCount(ByVal frameHeader As ICommonFrameHeader) As Int16
            Get
                Return frameHeader.InternalSampleCount And Not FrameTypeMask
            End Get
            Set(ByVal value As Int16)
                If value > MaximumSampleCount Then
                    Throw New OverflowException("Sample count value cannot exceed " & MaximumSampleCount)
                Else
                    frameHeader.InternalSampleCount = (frameHeader.InternalSampleCount And FrameTypeMask) Or value
                End If
            End Set
        End Property

        Public Shared Property IsFirstFrame(ByVal frameHeader As ICommonFrameHeader) As Boolean
            Get
                Return (frameHeader.InternalSampleCount And Bit12) = 0
            End Get
            Set(ByVal value As Boolean)
                If value Then
                    frameHeader.InternalSampleCount = frameHeader.InternalSampleCount And Not Bit12
                Else
                    frameHeader.InternalSampleCount = frameHeader.InternalSampleCount Or Bit12
                End If
            End Set
        End Property

        Public Shared Property IsLastFrame(ByVal frameHeader As ICommonFrameHeader) As Boolean
            Get
                Return (frameHeader.InternalSampleCount And Bit11) = 0
            End Get
            Set(ByVal value As Boolean)
                If value Then
                    frameHeader.InternalSampleCount = frameHeader.InternalSampleCount And Not Bit11
                Else
                    frameHeader.InternalSampleCount = frameHeader.InternalSampleCount Or Bit11
                End If
            End Set
        End Property

        Public Shared Property FrameCount(ByVal frameHeader As ICommonFrameHeader) As Int16
            Get
                Return frameHeader.InternalSampleCount And FrameCountMask
            End Get
            Set(ByVal value As Int16)
                If value > MaximumFrameCount Then
                    Throw New OverflowException("Frame count value cannot exceed " & MaximumFrameCount)
                Else
                    frameHeader.InternalSampleCount = (frameHeader.InternalSampleCount And Not FrameCountMask) Or value
                End If
            End Set
        End Property

        Public Shared Property FrameLength(ByVal frameHeader As ICommonFrameHeader) As Int16
            Get
                Return frameHeader.InternalStatusFlags And FrameLengthMask
            End Get
            Set(ByVal value As Int16)
                If value > MaximumFrameLength Then
                    Throw New OverflowException("Frame length value cannot exceed " & MaximumFrameLength)
                Else
                    frameHeader.InternalStatusFlags = (frameHeader.InternalStatusFlags And Not FrameLengthMask) Or value
                End If
            End Set
        End Property

        Public Shared Property DataLength(ByVal frameHeader As ICommonFrameHeader) As Int16
            Get
                ' Data length will be frame length minus common header length minus crc16
                Return FrameLength(frameHeader) - BinaryLength - 2
            End Get
            Set(ByVal value As Int16)
                If value > MaximumDataLength Then
                    Throw New OverflowException("Data length value cannot exceed " & MaximumDataLength)
                Else
                    FrameLength(frameHeader) = value + BinaryLength + 2
                End If
            End Set
        End Property

        Public Shared Property SynchronizationIsValid(ByVal frameHeader As ICommonFrameHeader) As Boolean
            Get
                Return (frameHeader.InternalStatusFlags And Bit15) = 0
            End Get
            Set(ByVal value As Boolean)
                If value Then
                    frameHeader.InternalStatusFlags = frameHeader.InternalStatusFlags And Not Bit15
                Else
                    frameHeader.InternalStatusFlags = frameHeader.InternalStatusFlags Or Bit15
                End If
            End Set
        End Property

        Public Shared Property DataIsValid(ByVal frameHeader As ICommonFrameHeader) As Boolean
            Get
                Return (frameHeader.InternalStatusFlags And Bit14) = 0
            End Get
            Set(ByVal value As Boolean)
                If value Then
                    frameHeader.InternalStatusFlags = frameHeader.InternalStatusFlags And Not Bit14
                Else
                    frameHeader.InternalStatusFlags = frameHeader.InternalStatusFlags Or Bit14
                End If
            End Set
        End Property

        Public Shared Property TriggerStatus(ByVal frameHeader As ICommonFrameHeader) As TriggerStatus
            Get
                Return frameHeader.InternalStatusFlags And TriggerMask
            End Get
            Set(ByVal value As TriggerStatus)
                frameHeader.InternalStatusFlags = (frameHeader.InternalStatusFlags And Not TriggerMask) Or value
            End Set
        End Property

    End Class

End Namespace

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/Ieee1344/CommonFrameHeader.cs
//*******************************************************************************************************
//  CommonFrameHeader.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  11/12/2004 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Security.Permissions;
using TVA;
using TVA.IO.Checksums;
using TVA.Parsing;

namespace PhasorProtocols.Ieee1344
{
    /// <summary>
    /// Represents the common header for all IEEE 1344 frames of data.
    /// </summary>
    [Serializable()]
    public class CommonFrameHeader : ICommonHeader<FrameType>, ISerializable
    {
        #region [ Members ]

        // Constants

        /// <summary>
        /// Total fixed length of <see cref="CommonFrameHeader"/>.
        /// </summary>
        public const ushort FixedLength = 6;

        // Fields
        private FrameImageCollector m_frameImages;
        private ushort m_sampleCount;
        private ushort m_statusFlags;
        private Ticks m_timestamp;
        private IChannelParsingState m_state;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="CommonFrameHeader"/> from specified parameters.
        /// </summary>
        /// <param name="typeID">The IEEE 1344 specific frame type of this frame.</param>
        /// <param name="timestamp">The timestamp of this frame.</param>
        public CommonFrameHeader(FrameType typeID, Ticks timestamp)
        {
            m_timestamp = timestamp;
            TypeID = typeID;
            FrameCount = 1;
            IsFirstFrame = true;
            IsLastFrame = true;
        }

        /// <summary>
        /// Creates a new <see cref="CommonFrameHeader"/> from given <paramref name="binaryImage"/>.
        /// </summary>
        /// <param name="configurationFrame">IEEE 1344 <see cref="ConfigurationFrame"/> if already parsed.</param>
        /// <param name="binaryImage">Buffer that contains data to parse.</param>
        /// <param name="startIndex">Start index into buffer where valid data begins.</param>
        public CommonFrameHeader(ConfigurationFrame configurationFrame, byte[] binaryImage, int startIndex)
        {
            uint secondOfCentury = EndianOrder.BigEndian.ToUInt32(binaryImage, startIndex);
            m_sampleCount = EndianOrder.BigEndian.ToUInt16(binaryImage, startIndex + 4);

            // We go ahead and pre-grab cell's status flags so we can determine framelength - we
            // leave startindex at 6 so that cell will be able to parse flags as needed - note
            // this increases needed common frame header size by 2 (i.e., BinaryLength + 2)
            m_statusFlags = EndianOrder.BigEndian.ToUInt16(binaryImage, startIndex + FixedLength);

            // NTP timestamps based on NtpTimeTag class are designed to work for dates between
            // 1968-01-20 and 2104-02-26 based on recommended bit interpretation in RFC-2030.
            NtpTimeTag timetag = new NtpTimeTag(secondOfCentury, 0);

            // Data frames have subsecond time information, so we add this fraction of time to current seconds value
            if (TypeID == Ieee1344.FrameType.DataFrame && configurationFrame != null)
                timetag.Value += SampleCount / Math.Truncate((double)Common.MaximumSampleCount / (double)configurationFrame.Period) / (double)configurationFrame.FrameRate;

            // Cache timestamp value
            m_timestamp = timetag.ToDateTime().Ticks;
        }

        /// <summary>
        /// Creates a new <see cref="CommonFrameHeader"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected CommonFrameHeader(SerializationInfo info, StreamingContext context)
        {
            // Deserialize common frame header
            m_sampleCount = info.GetUInt16("sampleCount");
            m_statusFlags = info.GetUInt16("statusFlags");
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets the timestamp of this frame in NTP format.
        /// </summary>
        public NtpTimeTag TimeTag
        {
            get
            {
                return new NtpTimeTag(m_timestamp);
            }
        }

        /// <summary>
        /// Gets or sets timestamp of this <see cref="CommonFrameHeader"/>.
        /// </summary>
        public Ticks Timestamp
        {
            get
            {
                return m_timestamp;
            }
            set
            {
                m_timestamp = value;
            }
        }

        /// <summary>
        /// Gets or sets the IEEE 1344 specific frame type of this frame.
        /// </summary>
        /// <remarks>
        /// <para>
        /// This returns the protocol specific frame classification which uniquely identifies the frame type.
        /// </para>
        /// <para>
        /// This is the <see cref="ICommonHeader{TTypeIdentifier}.TypeID"/> implementation.
        /// </para>
        /// </remarks>
        public FrameType TypeID
        {
            get
            {
                return (FrameType)(m_sampleCount & Common.FrameTypeMask);
            }
            set
            {
                m_sampleCount = (ushort)((m_sampleCount & ~Common.FrameTypeMask) | (ushort)value);
            }
        }

        /// <summary>
        /// Gets or sets flag from that determines if this is the first frame image in a series of images representing an entire frame.
        /// </summary>
        public bool IsFirstFrame
        {
            get
            {
                return (m_sampleCount & (ushort)Bits.Bit12) == 0;
            }
            set
            {
                if (value)
                    m_sampleCount = (ushort)(m_sampleCount & ~(ushort)Bits.Bit12);
                else
                    m_sampleCount = (ushort)(m_sampleCount | (ushort)Bits.Bit12);
            }
        }

        /// <summary>
        /// Gets or sets flag that determines if this is the last frame image in a series of images representing an entire frame.
        /// </summary>
        public bool IsLastFrame
        {
            get
            {
                return (m_sampleCount & (ushort)Bits.Bit11) == 0;
            }
            set
            {
                if (value)
                    m_sampleCount = (ushort)(m_sampleCount & ~(ushort)Bits.Bit11);
                else
                    m_sampleCount = (ushort)(m_sampleCount | (ushort)Bits.Bit11);
            }
        }

        /// <summary>
        /// Gets or sets the total frame count.
        /// </summary>
        public ushort FrameCount
        {
            get
            {
                return (ushort)(m_sampleCount & Common.FrameCountMask);
            }
            set
            {
                if (value > Common.MaximumFrameCount)
                    throw new OverflowException("Frame count value cannot exceed " + Common.MaximumFrameCount);
                else
                    m_sampleCount = (ushort)((m_sampleCount & ~Common.FrameCountMask) | value);
            }
        }
        
        /// <summary>
        /// Gets or sets the parsing state for the <see cref="CommonFrameHeader"/> object.
        /// </summary>
        public IChannelParsingState State
        {
            get
            {
                return m_state;
            }
            set
            {
                m_state = value;
            }
        }

        // Gets or sets any additional state information - satifies ICommonHeader<FrameType>.State interface property
        object ICommonHeader<FrameType>.State
        {
            get
            {
                return m_state;
            }
            set
            {
                m_state = value as IChannelParsingState;
            }
        }

        /// <summary>
        /// Gets the fundamental frame type of this frame.
        /// </summary>
        /// <remarks>
        /// Frames are generally classified as data, configuration or header frames. This returns the general frame classification.
        /// </remarks>
        public FundamentalFrameType FrameType
        {
            get
            {
                // Translate IEEE 1344 specific frame type to fundamental frame type
                switch (TypeID)
                {
                    case Ieee1344.FrameType.DataFrame:
                        return FundamentalFrameType.DataFrame;
                    case Ieee1344.FrameType.ConfigurationFrame:
                        return FundamentalFrameType.ConfigurationFrame;
                    case Ieee1344.FrameType.HeaderFrame:
                        return FundamentalFrameType.HeaderFrame;
                    default:
                        return FundamentalFrameType.Undetermined;
                }
            }
        }

        /// <summary>
        /// Gets or sets reference to a <see cref="FrameImageCollector"/> associated the IEEE 1344 frame, if any.
        /// </summary>
        public FrameImageCollector FrameImages
        {
            get
            {
                return m_frameImages;
            }
            set
            {
                m_frameImages = value;
            }
        }

        /// <summary>
        /// Gets or sets the entire length of the IEEE 1344 frame.
        /// </summary>
        public ushort FrameLength
        {
            get
            {
                return (ushort)(m_statusFlags & Common.FrameLengthMask);
            }
            set
            {
                if (value > Common.MaximumFrameLength)
                    throw new OverflowException("Frame length value cannot exceed " + Common.MaximumFrameLength);
                else
                    m_statusFlags = (ushort)((m_statusFlags & ~Common.FrameLengthMask) | value);
            }
        }

        /// <summary>
        /// Gets or sets the length of the data in the IEEE 1344 frame (i.e., the <see cref="FrameLength"/> minus the header length and checksum: <see cref="FrameLength"/> - 8).
        /// </summary>
        public ushort DataLength
        {
            get
            {
                // Data length will be frame length minus common header length minus crc16
                return (ushort)(FrameLength - FixedLength - 2);
            }
            set
            {
                if (value > Common.MaximumDataLength)
                    throw new OverflowException("Data length value cannot exceed " + Common.MaximumDataLength);
                else
                    FrameLength = (ushort)(value + FixedLength + 2);
            }
        }

        /// <summary>
        /// Gets or sets the sample number (i.e., frame count) of this frame.
        /// </summary>
        public ushort SampleCount
        {
            get
            {
                return (ushort)(m_sampleCount & ~Common.FrameTypeMask);
            }
            set
            {
                if (value > Common.MaximumSampleCount)
                    throw new OverflowException("Sample count value cannot exceed " + Common.MaximumSampleCount);
                else
                    m_sampleCount = (ushort)((m_sampleCount & Common.FrameTypeMask) | value);
            }
        }

        /// <summary>
        /// Gets the binary image of the common header portion of this frame.
        /// </summary>
        public byte[] BinaryImage
        {
            get
            {
                byte[] buffer = new byte[FixedLength];

                // Make sure frame length gets included in status flags for generated binary image
                FrameLength = (ushort)FixedLength;

                EndianOrder.BigEndian.CopyBytes((uint)TimeTag.Value, buffer, 0);
                EndianOrder.BigEndian.CopyBytes(m_sampleCount, buffer, 4);

                return buffer;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Appends header specific attributes to <paramref name="attributes"/> dictionary.
        /// </summary>
        /// <param name="attributes">Dictionary to append header specific attributes to.</param>
        internal void AppendHeaderAttributes(Dictionary<string, string> attributes)
        {
            attributes.Add("Frame Type", (ushort)TypeID + ": " + TypeID);

            if (FrameImages != null)
            {
                attributes.Add("Frame Length", FrameImages.BinaryLength.ToString());
                attributes.Add("Frame Images", FrameImages.Count.ToString());
            }
            else
            {
                attributes.Add("Frame Length", FrameLength.ToString());
                attributes.Add("Frame Images", "0");
            }
            
            attributes.Add("Frame Count", FrameCount.ToString());
            attributes.Add("Sample Count", m_sampleCount.ToString());
            attributes.Add("Status Flags", m_statusFlags.ToString());
            attributes.Add("Is First Frame", IsFirstFrame.ToString());
            attributes.Add("Is Last Frame", IsLastFrame.ToString());
        }

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        [SecurityPermission(SecurityAction.LinkDemand, Flags = SecurityPermissionFlag.SerializationFormatter)]
        public virtual void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            // Serialize unique common frame header values
            info.AddValue("sampleCount", m_sampleCount);
            info.AddValue("statusFlags", m_statusFlags);
        }

        #endregion

        #region [ Static ]

        // Static Methods

        /// <summary>
        /// Validates the CRC-CCITT for the specified IEEE 1344 buffer.
        /// </summary>
        public static bool ChecksumIsValid(byte[] buffer, int startIndex, int length)
        {
            int sumLength = length - 2;
            return EndianOrder.BigEndian.ToUInt16(buffer, startIndex + sumLength) == buffer.CrcCCITTChecksum(startIndex, sumLength);
        }

        #endregion       
    }
}